### PR TITLE
Properly surface error

### PR DESCRIPTION
**Pull request by bangtoven**

### CODE REVIEW

**Comment by Jonathanmarsh4** [comment]
https://www.google.co.uk/search?q=dash%3AXcxjrkwwNayx51pujr8SYyvmci5qXeBctg%3Famount%3D100000%26currency%3DUSD%26local%3D9562360.00&ie=UTF-8&oe=UTF-8&hl=en-gb&client=safari

**Review by erin-at-work — APPROVED** [review]
🚀

File: src/relay/WalletSDKRelay.ts [hunk 1]
```diff
@@ -1046,14 +1046,11 @@ export class WalletSDKRelay extends WalletSDKRelayAbstract {
...
-          if (error) {
+          if (error && (error instanceof ErrorResponse)) {
             this.handleWeb3ResponseMessage(
               Web3ResponseMessage({
                 id,
-                response: ErrorResponse(
-                  Web3Method.switchEthereumChain,
-                  error.message
-                )
+                response: (error as ErrorResponse)
```
Comment (ljharb) [inline src/relay/WalletSDKRelay.ts:1053]: is this typecast still needed? i'd expect TS to be able to type-narrow after the instanceof check above.

File: src/relay/WalletSDKRelay.ts [hunk 2]
```diff
@@ -1046,14 +1046,11 @@ export class WalletSDKRelay extends WalletSDKRelayAbstract {
         })
 
         const _cancel = (error?: Error) => {
-          if (error) {
+          if (error && (error instanceof ErrorResponse)) {
```
Comment (ljharb) [inline src/relay/WalletSDKRelay.ts:1049]:
```suggestion
          if (error instanceof ErrorResponse) {
```
a falsy value won't be an instanceof anything

File: src/relay/WalletSDKRelay.ts (hunk 1 above)
Comment (bangtoven) [reply src/relay/WalletSDKRelay.ts:1053]: yeah i hoped so, but for some reason it still thinks that's just `Error`

File: src/provider/WalletSDKUI.ts [hunk 3]
```diff
@@ -198,8 +194,7 @@ export class WalletSDKUI implements WalletUI {
     return false
   }
 
-  // @ts-ignore
-  inlineAddEthereumChain(chainId: string): boolean {
+  inlineAddEthereumChain(_chainId: string): boolean {
```
Comment (jeongmincho) [inline src/provider/WalletSDKUI.ts:197]: will this be updated?

File: src/relay/Web3Response.ts [hunk 4]
```diff
@@ -10,18 +10,23 @@ interface BaseWeb3Response<Result> {
...
   errorMessage: string
+  name: string
+  message: string
+
+  constructor(method: Web3Method, errorMessage: string, errorCode?: number) {
+    this.method = method
+    this.errorMessage = errorMessage
+    this.errorCode = errorCode
+    this.name = method
+    this.message = errorMessage
```
Comment (jeongmincho) [inline src/relay/Web3Response.ts:25]: q - is this to be compatible with how extension works with Error objects?

File: src/relay/Web3Response.ts [hunk 5]
```diff
@@ -10,18 +10,23 @@ interface BaseWeb3Response<Result> {
...
 }
 
-export function ErrorResponse(
-  method: Web3Method,
-  errorMessage: string,
-  errorCode?: number
-): ErrorResponse {
-  return { method, errorMessage, errorCode }
-}
+export const USER_REJECTED_REQUEST_ERROR = new Error("User rejected request")
```
Comment (jeongmincho) [inline src/relay/Web3Response.ts:29]: q - do we need to all caps this?

File: src/relay/WalletSDKRelay.ts [hunk 6]
```diff
@@ -537,12 +538,15 @@ export class WalletSDKRelay extends WalletSDKRelayAbstract {
     let hideSnackbarItem: (() => void) | null = null
     const id = randomBytesHex(8)
 
-    const cancel = () => {
+    const cancel = (error?: Error) => {
```
[the remark this answers is not in the document]
Comment (bangtoven) [reply src/relay/WalletSDKRelay.ts:541]: from ciphercore

File: src/relay/Web3Response.ts (hunk 5 above)
Comment (bangtoven) [reply src/relay/Web3Response.ts:29]: following the pattern on the extension, but open to suggestions.

**Review by erin-at-work — APPROVED** [review]

**Pull request closed without merging**